packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -606,6 +611,188 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     return workInProgress.child;
   }
 
+  function pushContextProvider(workInProgress) {
+    const context: ReactContext<any> = workInProgress.type;
+    // Store a reference to the previous provider
+    // TODO: Only need to do this on mount
+    workInProgress.stateNode = context.lastProvider;
+    // Push this onto the list of providers. We'll pop in the complete phase.
+    context.lastProvider = workInProgress;
+  }
+
+  function propagateContextChange<V>(
+    workInProgress: Fiber,
+    context: ReactContext<V>,
+    renderExpirationTime: ExpirationTime,
+  ): void {
+    let fiber = workInProgress.child;
+    while (fiber !== null) {
+      let nextFiber;
+      // Visit this fiber.
+      switch (fiber.tag) {
+        case ConsumerComponent:
+          // Check if the context matches.
+          if (fiber.type === context) {
+            // Update the expiration time of all the ancestors, including
+            // the alternates.
+            let node = fiber;
+            while (node !== null) {
+              const alternate = node.alternate;
+              if (
+                node.expirationTime === NoWork ||
+                node.expirationTime > renderExpirationTime
+              ) {
+                node.expirationTime = renderExpirationTime;
+                if (
+                  alternate !== null &&
+                  (alternate.expirationTime === NoWork ||
+                    alternate.expirationTime > renderExpirationTime)
+                ) {
+                  alternate.expirationTime = renderExpirationTime;
+                }
+              } else if (
+                alternate !== null &&
+                (alternate.expirationTime === NoWork ||
+                  alternate.expirationTime > renderExpirationTime)
+              ) {
+                alternate.expirationTime = renderExpirationTime;
+              } else {
+                // Neither alternate was updated, which means the rest of the
+                // ancestor path already has sufficient priority.
+                break;
+              }
+              node = node.return;
+            }
+            // Don't scan deeper than a matching consumer. When we render the
+            // consumer, we'll continue scanning from that point. This way the
+            // scanning work is time-sliced.
```

Comment:
does this imply that the internal subscriptions from react-redux (https://github.com/reactjs/react-redux/blob/master/src/utils/Subscription.js#L58) will be handled by React itself then?

Reply:
That's the plan